src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -1017,7 +1022,18 @@ public override void Write(byte[] buffer, int offset, int count)
                     int bytesRead = _deflateStream._inflater.Inflate(new Span<byte>(_arrayPoolBuffer));
                     if (bytesRead > 0)
                     {
+                        bool greaterThanUncompressedSize = false;
```

Comment:
I don't think solely limiting here is sufficient.  I think all the Read members on the stream itself will not go through this codepath, but should also limit.  You could either duplicate the limiting logic there or push it down to the inflater.  Did you consider these options?

Reply:
If i understood you correctly you mean about WriteAsync(...) method? That was actually my next question to ask from you, like when/how they are called and should i limit their read too, i guess we should limit it too in which case need to have other solution instead of updating uncompressed size with each byteRead

Reply:
Also there is another deflator DeflateManagedStream need to be covered

Reply:
> I see, you're modifying the underlying compressed size. I'd avoid that and instead do the math in the check and make the underlying field readonly. Or consider renaming it to "remainingUncompressedBytes". That will make it clearer. I personally prefer the former since it avoids new mutable state on the object.

in case we need to limit async writes, would follow your "former" suggestion

Reply:
> you mean about WriteAsync(...) method?

More than that. 

You're trying to make DeflateStream only read and copy up to a certain uncompressed size.

To do that you need to make sure that all Read and Copy methods will either not proceed past the given uncompressed size, or throw.  All of the places I see potentially reading past the compressed size are here:
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L230
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L264
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L398
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L471
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L939
... probably more.  Looks to me like every time Inflate is called you need to check to make sure it wouldn't exceed the uncompressed size.

Reply:
sure, will check them

Reply:
@ericstj I have moved the logic to inflate method and added more tests please re review